core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -128,6 +137,66 @@ public static <T extends ScanTask> CloseableIterable<ScanTaskGroup<T>> planTaskG
         combinedTasks -> new BaseScanTaskGroup<>(mergeTasks(combinedTasks)));
   }
 
+  @SuppressWarnings("unchecked")
+  public static <T extends PartitionScanTask> List<ScanTaskGroup<T>> planTaskGroups(
+      List<T> tasks,
+      long splitSize,
+      int lookback,
+      long openFileCost,
+      Types.StructType projectedPartitionType) {
+
+    Preconditions.checkArgument(splitSize > 0, "Invalid split size (negative or 0): %s", splitSize);
+    Preconditions.checkArgument(
+        lookback > 0, "Invalid split planning lookback (negative or 0): %s", lookback);
+    Preconditions.checkArgument(
+        openFileCost >= 0, "Invalid file open cost (negative): %s", openFileCost);
+
+    List<T> splitTasks = Lists.newArrayList();
```

Comment:
What about skipping this temporary list and simply checking if a task is splittable in the for loop below? All of this will be held in the driver memory and I am a bit paranoid.

```
if (task instanceof SplittableScanTask<?>) {
  SplittableScanTask<? extends T> splittableTask = (SplittableScanTask<? extends T>) task;
  Iterables.addAll(sameKeyTasks, splittableTask.split(splitSize));
} else {
  sameKeyTasks.add(task);
}
```

Reply:
Hmm I'm not sure how much this could help to be honest. It seems we mainly want to avoid materializing `splittableTask.split(splitSize)` here, but it's going to be done anyway immediately later.

The list here only keeps references to the tasks which should be pretty tiny comparing to the space taken by the tasks themselves.

Reply:
I agree it will be only about the extra spaces occupied by temp list slots. Up to you. I am okay either way.